README.rst
=====
Bravo
=====

Bravo is a elegant, speedy, and extensible implementation of the Minecraft
Alpha/Beta/"Modern" protocol. Only the server side is implemented. Bravo also
has a few tools useful for examining the wire protocols and disk formats in
Minecraft.

Features
========

Standard Features
-----------------

 * Console
 * Login and handshake
 * Geometry ("chunk") transfer
 * Location updates
 * Passage of time (day/night)
 * Block construction and deconstruction
 * Entities

   * Players
   * Pickups
   * Tiles

     * Chests
     * Furnaces
     * Signs

 * Lighting
 * Save controls
 * Server-side inventories

Extended Features
-----------------

 * Pluggable architecture

   * Commands

     * Inventory control
     * Teleports
     * Time of day

   * Terrain generation

     * Erosion
     * Simplex noise, 2D and 3D
     * Water table

   * Seasons

     * Spring
     * Winter

   * Physics

     * Sand, gravel
     * Water, lava
     * Redstone

 * Chat commands
 * IP ban list

Installing
==========

Bravo currently requires Python 2.6 or any newer Python 2.x. It is known to
work on CPython and PyPy.

Bravo ships with a standard setup.py. You will need setuptools/distribute, but
most distributions already provide it for you. Bravo depends on the following
external libraries from PyPI:

 * construct, version 2.03 or later
 * Twisted, version 11.0 or later

If installing modular Twisted, Twisted Conch is required.

For IRC support, Twisted Words is required; it is usually called
python-twisted-words or twisted-words in package managers.

For web service support, Twisted Web must be installed; it is generally called
python-twisted-web or twisted-web.

Windows
-------

No installation required, a standalone executable is available at:

::

 http://bravoserver.org/downloads.html


Debian & Ubuntu
---------------

Debian and its derivatives, like Ubuntu, have Twisted in their
package managers.

::

 $ sudo aptitude install python-twisted

If you are tight on space, you can install only part of Twisted.

::

 $ sudo aptitude install python-twisted-core python-twisted-bin python-twisted-conch

A Note about Ubuntu
^^^^^^^^^^^^^^^^^^^

If you are using Ubuntu 10.04 LTS, you will need a more recent Twisted than
Ubuntu provides. There is a PPA at
http://launchpad.net/~twisted-dev/+archive/ppa which provides recent versions
of all Twisted packages.

Fedora
------

Twisted can be installed from the standard Fedora repository.

::

 $ sudo yum install python-twisted python-twisted-conch

Gentoo
------

Gentoo does carry a Construct new enough for Bravo, but it does have to be
unmasked.

::

 # emerge twisted twisted-conch

LFS/Virtualenv/Standalone
-------------------------

If, for some reason, you are installing to a very raw or unmanaged place, and
you want to ensure that everything is built from the latest source available
on PyPI, we highly recommend pip for installing Bravo, since it handles all
dependencies for you.

::

 $ pip install Bravo

Bravo can also optionally use Ampoule to offload some of its inner
calculations to a separate process, improving server response times. Ampoule
will be automatically detected and is completely optional.

::

 $ pip install ampoule

Running
=======

Bravo includes a twistd plugin, so it's quite easy to run. Just copy
bravo.ini.example to bravo.ini, and put it in one of these locations:

 * /etc/bravo/
 * ~/.bravo/
 * Your working directory

And then run the Twisted plugin:

::

 $ twistd -n bravo

Contributing
============

Contributing is easy! Just send me your code. Diffs are appreciated, in git
format; Github pull requests are excellent.

Things to consider:

 * I will be rather merciless about your code during review, especially if it
   adds lots of new features.
 * Some things are better off outside of the main tree, especially if they are
   moving very fast compared to Bravo itself.
 * Unit tests are necessary for new code, especially feature-laden code. If
   your code is absolutely not testable, it's not really going to be very fun
   to maintain. See the above point.
 * Bravo is MIT/X11. Your contributions will be under that same license. If
   this isn't acceptable, then your code cannot be merged. This is really the
   only hard condition.

FAQ
===

The FAQ moved to the docs; see docs/faq.rst, or more usefully,
http://bravo.readthedocs.org/en/latest/faq.html for an HTML version.

License
=======

Bravo is MIT/X11-licensed. See the LICENSE file for the actual text of the license.


CHANGELOG.rst
2.0
===

Compatibility
-------------

* Bravo is now fully compatible with PyPy. Accordingly, PyPy is now supported.
* Bravo no longer uses Exocet.

  * All hooking plugins that expect to have access to a factory must now have
    an ``__init__()`` which takes a ``factory`` keyword argument.
* Recipes and seasons are no longer pluggable.
* Breakage related to the year-long hiatus has been largely fixed.

  * The 1.4.x protocol is now supported, without server-side encryption.
  * The Alpha and Beta serializers have been removed in favor of a single Anvil
    serializer. (Contact the developers if you need Alpha/Beta support!)
  * Online authentication was broken and has been disabled.
  * Ascension and the ``/ascend`` command are unbroken. (#390)
  * Chat commands are enabled again. (#391)
  * Building on snow has been moved from pluggable to core functionality.
    (#410)
* ``Location``'s API has been cleaned up and made internally consistent.
  (#368, #373)
* Server configuration must now be specified with ``-c`` on the command line
  when starting Bravo. (#386)
* Water level has been lowered a couple blocks, for consistency with Vanilla
  terrain generation. (#411)
* Bravo now boasts support for the Win32 operating system platform, with build
  scripts in-tree. (#412)

Features
--------

* Seeds can be set for worlds in configuration files (#239)
* Plugins can be grouped into packs, and there is a pack for Beta (#380)
* Jack-o-lanterns are available in creative (#392)
* Emerald blocks, ores, and items are available in creative (#394)

Bugfixes
--------

* Fixed client being kicked when shooting arrows (#378)
* Removed magic numbers in painting directions (#379)
* Allowed placement of certain creative blocks (#392)
* Fixed incorrect packet formatting in Player entities (#396)
* Colorized nicks in server join messages (#399)
* Re-enabled track orientation (#404)

1.9
===

Compatibility
-------------

* bravo.config.configuration no longer exists and there is no longer a global
  configuration object; use factory.config or protocol.config to obtain the
  current configuration. (#376)
* Serializer refactoring

  * Serializers now start when their ``connect()`` method is called, rather
    than on initialization
  * Serializers may raise ``SerializerReadException`` if asked to retrieve
    data which does not exist
  * Serializers may ``implements(ISerializer)`` directly
  * ``ISerializerFactory`` no longer exists
  * Serializers should no longer need to be parameterized, and will not
    receive configuration objects or worlds

Features
--------

* Enabled certain simple redstone circuitry; redstone is now a finished
  feature which can have issues filed for it, rather than a wishlist item
* Added blocks and crafting recipes for pistons and sticky pistons (#375)

Removals
--------

* Removed password authentication support

Bugfixes
--------

* Fixed sand not falling on top of snow when AlphaSandGravel is enabled (#298)
* Fixed ultra-large or bogus ping times causing server crashes in Python 2.7+
  (#363)
* Fixed player list not showing up in Notchian client

1.8.1
=====

Compatibility
-------------

* Syntax errors in plugins are now caught and logged

Features
--------

* Cleaned up furnaces to be slightly more efficient (#361)
* Added new blocks for 1.7/1.8 (#364)
* Sped up chunk lighting slightly
* Added redstone circuit for levers
* Rewrote redstone to be enabled for torches, plain blocks, and levers

Bugfixes
--------

* Fixed lighting not being regenerated after block changes (#132)
* Fixed blocks not being diggable in creative mode (#349)
* Fixed /time command failing when no seasons are enabled (#360)
* Fixed block representations not being printable

1.8
===

Compatibility
-------------

Blocks
^^^^^^

* Block drops now consider both the primary and secondary attributes of a drop
* Several blocks have been renamed to avoid block/item conflicts

  * Block "lapis-lazuli" to "lapis-lazuli-block"
  * Block "bed" to "bed-block" 
  * Block "diamond" to "diamond-block"
  * Block "wooden-door" to "wooden-door-block"
  * Block "iron-door" to "iron-door-block"
  * Block "sugar-cane" to "reed"
  * Block "cake" to "cake-block"

* "Light green" blocks are now "lime" (#344)

  * Block "light-green-wool" to "lime-wool"

Modules
^^^^^^^

* All Beta-related classes have been reconsolidated under the ``bravo.beta``
  namespace

Plugins
^^^^^^^

* Recipes have been completely reworked to carry their own crafting table
  logic; see ``bravo.beta.recipes`` for helper classes
* New simple hooks for taking action when windows are opened, clicked, and
  closed
* New simple hook for taking action at the beginning of a digging action
* Pre-build hooks have an altered signature for ``pre_build_hook()``, with a
  new return value and new semantics
* Post-build hooks explicitly have their input coordinates pre-adjusted

Features
--------

* Support for Beta protocol 17 (Beta 1.8.x) (#346)

  * Server polling and announce (#350)
  * Controls for factory modes: Creative and survival (#353)
  * In-game player lists (#358)

* Complete rework of inventory handling

  * Support for creative inventories
  * Support for chests (#256)
  * Support for furnaces (#261)

* Rework of recipes

  * Support for crafting ingredient-based recipes
  * Crafting dyes (#331)
  * Crafting with non-white wool (#336)

* Support for total and per-IP connection limits (#310)
* Connection timeouts for clients (#319)
* Support for placing beds (#255)
* Chunk feeding for automatons (#271)
* Worldmap web plugin scrolling support
* Documentation reorganization to aid new administrators and developers

Bugfixes
--------

* Fixed certain crafting operations crashing Beta 1.6+ clients (#302)
* Fixed sand and gravel not replacing water and lava when falling (#317)
* Fixed tracks not orienting when placed (#326)
* Fixed dropped items reappearing on reconnect (#330)
* Fixed lapis lazuli ore not dropping lapis lazuli dye (#357)

packets.py
import collections
import functools
import sys

from construct import Struct, Container, Embed
from construct import MetaArray, If
from construct import PascalString
from construct import UBInt8, UBInt16, UBInt32, UBInt64
from construct import SBInt16, SBInt32
from construct import BFloat32, BFloat64

from construct.core import ArrayError, FieldError

# Our tricky Java string decoder.
# Note that Java has a weird encoding for the NULL byte which we do not
# respect or honor since no client will generate it. Instead, we will get two
# NULL bytes in a row.
AlphaString = functools.partial(PascalString,
    length_field=UBInt16("length"),
    encoding="utf8")

flying = Struct("flying", UBInt8("flying"))
position = Struct("position",
    BFloat64("x"),
    BFloat64("y"),
    BFloat64("stance"),
    BFloat64("z")
)
look = Struct("look", BFloat32("rotation"), BFloat32("pitch"))

entity = Struct("entity", UBInt32("id"))
entity_position = Struct("entity_position",
    entity,
    UBInt8("x"),
    UBInt8("y"),
    UBInt8("z")
)
entity_look = Struct("entity_look",
    entity,
    UBInt8("rotation"),
    UBInt8("pitch")
)
entity_position_look = Struct("entity_position_look",
    entity,
    UBInt8("x"),
    UBInt8("y"),
    UBInt8("z"),
    UBInt8("rotation"),
    UBInt8("pitch")
)

items = collections.defaultdict(lambda: "unused")
items.update({
    4: "cobblestone",
    20: "glass",
})

packets = {
    0: Struct("ping"),
    1: Struct("login",
        UBInt32("protocol"),
        AlphaString("username"),
        AlphaString("unused"),
        UBInt64("unknown1"),
        UBInt8("unknown2"),
    ),
    2: Struct("handshake",
        AlphaString("username"),
    ),
    3: Struct("chat",
        AlphaString("message"),
    ),
    4: Struct("time",
        UBInt64("timestamp"),
    ),
    5: Struct("inventory",
        SBInt32("unknown1"),
        UBInt16("length"),
        MetaArray(lambda context: context["length"],
            Struct("items",
                SBInt16("id"),
                If(lambda context: context["id"] >= 0,
                    Embed(Struct("item_information",
                        UBInt8("count"),
                        UBInt16("damage"),
                    )),
                ),
            ),
        ),
    ),
    6: Struct("spawn",
        UBInt32("x"),
        UBInt32("y"),
        UBInt32("z"),
    ),
    10: flying,
    11: Struct("position", position, flying),
    12: Struct("look", look, flying),
    13: Struct("location", position, look, flying),
    14: Struct("digging",
        UBInt8("state"),
        SBInt32("x"),
        UBInt8("y"),
        SBInt32("z"),
        UBInt8("face"),
    ),
    15: Struct("build",
        UBInt16("block"),
        SBInt32("x"),
        UBInt8("y"),
        SBInt32("z"),
        UBInt8("face"),
    ),
    16: Struct("equip",
        UBInt32("entity"),
        UBInt16("item"),
    ),
    17: Struct("pickup",
        UBInt16("type"),
        UBInt8("quantity"),
        UBInt16("wear"),
    ),
    18: Struct("arm",
        entity,
        UBInt8("animate"),
    ),
    20: Struct("unknown1",
        UBInt32("unknown2"),
        AlphaString("unknown3"),
        UBInt32("unknown4"),
        UBInt32("unknown5"),
        UBInt32("unknown6"),
        UBInt8("unknown7"),
        UBInt8("unknown8"),
        UBInt16("unknown9"),
    ),
    21: Struct("pickup",
        entity,
        UBInt16("item"),
        UBInt8("count"),
        SBInt32("x"),
        SBInt32("y"),
        SBInt32("z"),
        UBInt8("yaw"),
        UBInt8("pitch"),
        UBInt8("roll"),
    ),
    22: Struct("collect",
        entity,
        UBInt32("destination"),
    ),
    23: Struct("unknown1",
        UBInt32("unknown"),
        UBInt8("unknown"),
        UBInt32("unknown"),
        UBInt32("unknown"),
        UBInt32("unknown"),
    ),
    24: Struct("unknown1",
        entity,
        UBInt8("unknown3"),
        SBInt32("x"),
        SBInt32("y"),
        SBInt32("z"),
        UBInt8("unknown7"),
        UBInt8("unknown8"),
    ),
    29: Struct("destroy",
        entity,
    ),
    30: entity,
    31: entity_position,
    32: entity_look,
    33: entity_position_look,
    34: Struct("unknown",
        UBInt32("unknown"),
        UBInt32("unknown"),
        UBInt32("unknown"),
        UBInt32("unknown"),
        UBInt8("unknown6"),
        UBInt8("unknown6"),
    ),
    50: Struct("chunk_enable",
        SBInt32("x"),
        SBInt32("z"),
        UBInt8("enabled"),
    ),
    51: Struct("chunk",
        SBInt32("x"),
        UBInt16("y"),
        SBInt32("z"),
        UBInt8("x_size"),
        UBInt8("y_size"),
        UBInt8("z_size"),
        PascalString("data", length_field=UBInt32("length")),
    ),
    52: Struct("batch",
        SBInt32("x"),
        SBInt32("z"),
        UBInt16("length"),
        MetaArray(lambda context: context["length"], UBInt16("coords")),
        MetaArray(lambda context: context["length"], UBInt8("types")),
        MetaArray(lambda context: context["length"], UBInt8("metadata")),
    ),
    53: Struct("block",
        SBInt32("x"),
        UBInt8("y"),
        SBInt32("z"),
        UBInt8("type"),
        UBInt8("meta"),
    ),
    59: Struct("tile_entity",
        SBInt32("x"),
        SBInt16("y"),
        SBInt32("z"),
        PascalString("nbt", length_field=UBInt16("length")),
    ),
    255: Struct("error",
        AlphaString("message"),
    ),
}

def parse_packets(bytestream):
    """
    Opportunistically parse out as many packets as possible from a raw
    bytestream.

    Returns a tuple containing a list of unpacked packet containers, and any
    leftover unparseable bytes.
    """

    l = []
    marker = 0

    while bytestream:
        header = ord(bytestream[0])

        if header in packets:
            parser = packets[header]
            try:
                container = parser.parse(bytestream[1:])
            except (ArrayError, FieldError):
                break
            except Exception, e:
                print type(e), e
                break

            # Reconstruct the packet and discard the data from the stream.
            # The extra one is for the header; we want to advance the stream
            # as atomically as possible.
            rebuilt = parser.build(container)
            length = 1 + len(rebuilt)
            bytestream = bytestream[length:]
            print "Parsed packet %d (#%d) @ %#x (%#x bytes)" % (
                header, len(l), marker, length)
            l.append((header, container))
            marker += length
        else:
            print "Couldn't decode packet %d @ %#x" % (header, marker)
            sys.exit()

    return l, bytestream

def make_packet(packet, **kwargs):
    """
    Constructs a packet bytestream from a packet header and payload.

    The payload should be passed as keyword arguments.
    """

    header = chr(packet)
    payload = packets[packet].build(Container(**kwargs))
    return header + payload

def make_error_packet(message):
    """
    Convenience method to generate an error packet bytestream.
    """

    return make_packet(255, message=message)

